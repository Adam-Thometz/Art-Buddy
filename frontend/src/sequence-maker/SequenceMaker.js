import React from "react";

import { useSelector } from "react-redux";

import Categories from "./categories/Categories";
import Options from "./options/Options";
import Sequence from "./sequence/Sequence";

const SequenceMaker = () => {
  const category = useSelector(state => state.sequenceMaker.category);
  return (
    <div className="SequenceMaker">
      <Categories />
      {category ? <Options /> : null}
      <Sequence />
    </div>
  );
};

export default SequenceMaker;