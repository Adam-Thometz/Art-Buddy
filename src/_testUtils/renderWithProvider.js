import { render } from '@testing-library/react';
import { Provider } from 'react-redux';
import { MemoryRouter } from 'react-router-dom';

// import renderWithRouter from './renderWithRouter';
import rootReducer from '../_redux/reducers/rootReducer';

function withProvider({ children, store }) {
  return (
    <Provider store={store}>
      <MemoryRouter>
        {children}
      </MemoryRouter>
    </Provider>
  )
};

function renderWithProvider(component, options) {
  const store = (options && options.store)
    ? options.store
    : rootReducer;
  return render(component, { wrapper: (args) =>
    withProvider({
      ...args,
      store
    }), ...options
  });
};

export default renderWithProvider;