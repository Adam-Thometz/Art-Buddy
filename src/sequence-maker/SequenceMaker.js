import React from "react";

import { useSelector, useDispatch } from "react-redux";
import { changeCategory, addToSequence } from "../_redux/actions";

import Sequence from "./sequence/Sequence";
import Dropdown from "../_components/dropdown/Dropdown";
import Icon from "../_components/icon/Icon";
import Options from "../_components/option/Options";

import categories from "./_utils/categories";
import soundInfo from "./_media/soundInfo";

const SequenceMaker = () => {
  const category = useSelector(state => state.sequenceMaker.category);
  const sequence = useSelector(state => state.sequenceMaker.sequence);
  const dispatch = useDispatch();

  const handleChangeCategory = e => {
    dispatch(changeCategory(e.target.value));
  }

  const handleAddToSequence = e => {
    if (sequence.some(block => !block)) {
      const sound = e.currentTarget.classList[1];
      dispatch(addToSequence(sound));
    };
  }

  const options = soundInfo[category];

  return (
    <div className="SequenceMaker">
      <Dropdown
        labelText="SELECT A SOUND CATEGORY"
        onChange={handleChangeCategory}
        options={categories}
      />
      {category ? (
        <Options>
          {Object.keys(options).map(option => (
            <Icon
              icon={options[option].image}
              size='150px'
              text={options[option].alt}
              additionalClass={option}
              onClick={handleAddToSequence}
            />
          ))}
        </Options>
      ) : null}
      <Sequence />
    </div>
  );
};

export default SequenceMaker;