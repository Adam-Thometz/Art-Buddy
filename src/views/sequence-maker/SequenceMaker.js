import { useContext, useEffect } from "react";
import useVisited from "hooks/visited/useVisited";
import { PlayContext } from "context/PlayContext";

import { useDispatch, useSelector } from "react-redux";
import { clearGame } from "store/sequence-maker/sequenceMakerReducer";
import { changeCurrGame } from "store/_general/generalReducer";

import WindowNavbar from "components/window-nav/WindowNavbar";
import Instructions from "components/instructions/Instructions";
import SequenceControls from "./controls/SequenceControls";
import SoundOptions from "./sound-options/SoundOptions";
import Sequence from "./sequence/Sequence";
import SequencePlayReset from "./play-reset/SequencePlayReset";

import { sequenceMaker } from "data/_activities/activityList";
import { Transport } from "tone";
import createSounds from "lib/sequence-maker/createSounds";

const SequenceMaker = () => {
  const [hasVisited, setHasVisited] = useVisited(sequenceMaker.lsKey);
  const { currGame } = useSelector((state) => state.general);
  const { volume } = useSelector((state) => state.settings);
  const { sequence } = useSelector((state) => state.sequenceMaker);
  const { setPlayFn } = useContext(PlayContext);
  const dispatch = useDispatch();

  useEffect(() => {
    dispatch(changeCurrGame(sequenceMaker));
    return () => {
      dispatch(clearGame());
      dispatch(changeCurrGame({}));
      Transport.stop();
    };
  }, [dispatch]);

  useEffect(() => {
    setPlayFn(() => createSounds(sequence, volume));
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [sequence, volume]);

  return (
    <>
      <WindowNavbar page={currGame.name} />
      {!hasVisited ? (
        <Instructions setVisited={setHasVisited} />
      ) : (
        <>
          <SequenceControls />
          <SoundOptions />
          <Sequence />
          <SequencePlayReset />
        </>
      )}
    </>
  );
};

export default SequenceMaker;
