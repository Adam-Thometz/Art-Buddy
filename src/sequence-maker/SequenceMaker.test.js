import React from "react";

import renderWithProvider from '../_testUtils/renderWithProvider';
import { screen } from "@testing-library/react";
import userEvent from "@testing-library/user-event";

import SequenceMaker from './SequenceMaker';

describe('SequenceMaker component', () => {
  it('renders without crashing', () => {
    renderWithProvider(<SequenceMaker />);
  });

  it('matches snapshot', () => {
    const { asFragment } = renderWithProvider(<SequenceMaker />);
    expect(asFragment()).toMatchSnapshot();
  });

  it('adds sounds to the sound options component', () => {
    renderWithProvider(<SequenceMaker />);
    const categoryDropdown = screen.getByText('SOUND CATEGORY');
    userEvent.hover(categoryDropdown);
    const option = screen.getByText('ANIMALS');
    userEvent.click(option);
    expect(screen.getByText('DOG')).toBeInTheDocument();
  });
});