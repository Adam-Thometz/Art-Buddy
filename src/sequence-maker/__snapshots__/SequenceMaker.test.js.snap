// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SequenceMaker component matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="SequenceMaker"
  >
    <p>
      HULLO IM SEQUENCE MAKER MAKE A SEQUENCE PLZ
    </p>
    <hr />
    <div
      class="SequenceControls"
    >
      <div
        class="Dropdown"
      >
        <span
          class="Dropdown-label"
        >
          SOUND CATEGORY 
          <img
            alt=""
            src="file"
          />
        </span>
        <div
          class="Dropdown-options"
        >
          <p
            class="Dropdown-option"
            id="bodySoundsContinuous"
          >
            BODY SOUNDS
          </p>
          <p
            class="Dropdown-option"
            id="bodySoundsOnce"
          >
            BODY SOUNDS (once)
          </p>
          <p
            class="Dropdown-option"
            id="animals"
          >
            ANIMALS
          </p>
        </div>
      </div>
      <div
        class="Dropdown"
      >
        <span
          class="Dropdown-label"
        >
          PITCH 
          <img
            alt=""
            src="file"
          />
        </span>
        <div
          class="Dropdown-options"
        >
          <p
            class="Dropdown-option"
            id="2"
          >
            LOW
          </p>
          <p
            class="Dropdown-option"
            id="3"
          >
            MEDIUM
          </p>
          <p
            class="Dropdown-option"
            id="4"
          >
            HIGH
          </p>
        </div>
      </div>
      <div
        class="Dropdown"
      >
        <span
          class="Dropdown-label"
        >
          DURATION 
          <img
            alt=""
            src="file"
          />
        </span>
        <div
          class="Dropdown-options"
        >
          <p
            class="Dropdown-option"
            id="1"
          >
            1 SECOND
          </p>
          <p
            class="Dropdown-option"
            id="2"
          >
            2 SECONDS
          </p>
          <p
            class="Dropdown-option"
            id="3"
          >
            3 SECONDS
          </p>
          <p
            class="Dropdown-option"
            id="4"
          >
            4 SECONDS
          </p>
          <p
            class="Dropdown-option"
            id="5"
          >
            5 SECONDS
          </p>
        </div>
      </div>
    </div>
    <div
      class="SoundOptions"
    />
    <div
      class="Sequence"
    >
      <div
        class="Sequence-play-options"
      >
        <div
          class="Icon"
        >
          <img
            alt=""
            src="file"
            style="height: 48px; width: 48px; cursor: pointer;"
          />
          <p
            style="font-size: 1rem;"
          >
            Play
          </p>
        </div>
        <div
          class="Icon"
        >
          <img
            alt=""
            src="file"
            style="height: 48px; width: 48px; cursor: pointer;"
          />
          <p
            style="font-size: 1rem;"
          >
            Play All
          </p>
        </div>
      </div>
      <div
        class="Sequence-display"
      >
        <div
          class="SequenceBlock"
          id="0"
          style="border-color: #89eb0a;"
        />
        <div
          class="SequenceBlock"
          id="1"
          style="border-color: #ffe81a;"
        />
        <div
          class="SequenceBlock"
          id="2"
          style="border-color: #fd4d3a;"
        />
        <div
          class="SequenceBlock"
          id="3"
          style="border-color: #4d4f5e;"
        />
      </div>
      <div
        class="Sequence-reset"
      >
        <div
          class="Icon"
        >
          <img
            alt=""
            style="height: 48px; width: 48px; cursor: pointer;"
          />
          <p
            style="font-size: 1rem;"
          >
            Reset Sequence
          </p>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
